portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -59,8 +59,25 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
     Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
     PreparedQuery results = datastore.prepare(query);
 
+    // Get number of comments to display
+    int numCommentsToDisplay;
+    try {
+      numCommentsToDisplay = Integer.parseInt(request.getParameter("limit"));
+      System.err.println(request.getParameter("limit"));
+    } catch (NumberFormatException e) {
+      System.err.println("Input for limit invalid. Displaying default number.");
+      numCommentsToDisplay = 5;
+    }
+
     ArrayList<Comment> comments = new ArrayList<>();
     for (Entity entity : results.asIterable()) {
+      // Check whether has reached comment display limit
+      if(numCommentsToDisplay == 0) {
+        break;
+      }
+      numCommentsToDisplay--;
```

Comment:
We can make this more efficient by making the query return a specific number of elements. Right now you are getting all the data from the database into the servlet. Imagine your database had millions of entries. Then in this case they would all be loaded into the servlet. Well actually I'm not sure how PreparedQuery.asIterable() is implemented but its good practice to make your query for any database do minimal work needed.

To fix this, look into `FetchOptions fetchOptions = FetchOptions.Builder.withLimit(maxComments);`